### PR TITLE
split read done channel from bufchan

### PR DESCRIPTION
**Pull request by alpha-baby**
### Issues associated with this PR

Your PR should present related issues you want to solve.

### Solutions
You should show your solutions about the issues in your PR, including the overall solutions, details and the changes. At this time, Chinese is allowed to describe these.

### UT result
Unit Test is needed if the code is changed, your unit test should cover boundary cases, corner cases, and some exceptional cases. And you need to show the UT result.

### Benchmark
If your code involves the processing of every request, you should give the Benchmark Result.

### Code Style
+ Make sure `Goimports` has run
+ Show `Golint` result

### CODE REVIEW

**Comment by alpha-baby** [comment]
buf 复现：

```log
--- FAIL: TestStreamConnectionDispatch (3.32s)
    stream_test.go:113: want body: hello world!. get:  world!hello
FAIL
```

我些的这个单元测试如果想要比较好的复现需要修改下 read函数：

```go
func (conn *streamConnection) Read(p []byte) (n int, err error) {
	// conn.bufChan <- nil Maybe caused panic error when connection closed.
	defer func() {
		if r := recover(); r != nil {
			log.DefaultLogger.Infof("[stream] [http] connection has closed. Connection = %d, Local Address = %+v, Remote Address = %+v, err = %+v",
				conn.conn.ID(), conn.conn.LocalAddr(), conn.conn.RemoteAddr(), r)
		}
	}()

	data, ok := <-conn.bufChan
	time.Sleep(time.Millisecond * 100) //tip 在这里添加上超时
	// Connection close
	if !ok {
		// Compatible with the fasthttp,
		// it should be set err = IO.EOF when the peer connection is closed.
		switch conn.resetReason {
		case types.UpstreamReset:
			err = io.EOF
		default:
			err = errConnClose
		}

		return
	}

	n = copy(p, data.Bytes())
	data.Drain(n)
	conn.bufChan <- nil
	return
}
```

`Dispatch`函数确实不会丢失数据，但是在`Dispatch`函数中会出现长时间的自旋，然后导致数据不一致

如果想要看到自旋的问题再到此函数中加入日志打印就可以：

```go
func (sc *streamConnection) Dispatch(buffer buffer.IoBuffer) {
	defer func() {
		if r := recover(); r != nil {
			log.DefaultLogger.Errorf("[stream] [http] connection has closed. Connection = %d, Local Address = %+v, Remote Address = %+v, err = %+v",
				sc.conn.ID(), sc.conn.LocalAddr(), sc.conn.RemoteAddr(), r)
		}
	}()

	for buffer.Len() > 0 {
		sc.bufChan <- buffer
		<-sc.bufChan
		log.DefaultLogger.Infof("len: %v", buffer.Len())
	}
	log.DefaultLogger.Infof("exit: %v", buffer.String())
}
```

最新 commit 修此问题，unit test 结果如下：

```log
=== RUN   TestStreamConnectionDispatch
2020-11-13 18:30:35,196 [INFO] len: 0
2020-11-13 18:30:35,196 [INFO] exit: 
2020-11-13 18:30:35,297 [INFO] len: 0
2020-11-13 18:30:35,297 [INFO] exit: 
--- PASS: TestStreamConnectionDispatch (3.31s)
2020-11-13 18:30:35,398 [INFO] len: 0
2020-11-13 18:30:35,398 [INFO] exit: 
PASS
```

**Comment by alpha-baby** [comment]
例如在我本地出现的自旋，输出的日志如下：

```log
2020-11-13 18:12:43,970 [INFO] len: 117
2020-11-13 18:12:43,970 [INFO] len: 5
2020-11-13 18:12:44,421 [INFO] len: 5
2020-11-13 18:12:44,803 [INFO] len: 117
2020-11-13 18:12:44,803 [INFO] len: 7
2020-11-13 18:12:45,196 [INFO] len: 5
2020-11-13 18:12:45,738 [INFO] len: 0
2020-11-13 18:12:45,738 [INFO] exit: 
2020-11-13 18:12:46,119 [INFO] len: 5
2020-11-13 18:12:46,467 [INFO] len: 7
2020-11-13 18:12:46,785 [INFO] len: 5
2020-11-13 18:12:47,131 [INFO] len: 7
2020-11-13 18:12:47,131 [INFO] len: 0
2020-11-13 18:12:47,131 [INFO] exit: 
2020-11-13 18:12:48,420 [INFO] len: 0
2020-11-13 18:12:48,420 [INFO] exit: 
// ...
// ....
```

**Comment by taoyuanyuan** [comment]
@alpha-baby 你的test是不对的，一个连接同时最多只会一个协程调用Dispatch，不会并发的。

**Comment by alpha-baby** [comment]
是因为在http1.X 协议汇总，一个TCP连接在同一个时间只会发送一个请求，是这样理解吗？

**Comment by alpha-baby** [comment]
我又看看下代码，确实是这样的，
如果是这样的话，我觉得我这样修改也可以增强代码的可读性和健壮性，
如果你们觉得没得必要的话，就把这个PR关掉就行了
@taoyuanyuan

**Comment by taoyuanyuan** [comment]
@cch123  PTAL

**Comment by cch123** [comment]
感觉可以，需要试一下 http 的性能有没有影响，用 example 里的 demo 和 wrk 简单测一下 post 上来就行，
如果差不多的话下版本合并~

**Comment by alpha-baby** [comment]
> 感觉可以，需要试一下 http 的性能有没有影响，用 example 里的 demo 和 wrk 简单测一下 post 上来就行，
> 如果差不多的话下版本合并~

test environment： MacBook Pro (16-inch, 2019) 2.6 GHz 六核Intel Core i7 16 GB 2667 MHz DDR4

start http-sample: `GOMAXPROCS=4 go run ./examples/codes/http-sample/server.go`

**Comment by alpha-baby** [comment]
感觉可以，需要试一下 http 的性能有没有影响，用 example 里的 demo 和 wrk 简单测一下 post 上来就行，
如果差不多的话下版本合并~

test environment： MacBook Pro (16-inch, 2019) 2.6 GHz 六核Intel Core i7 16 GB 2667 MHz DDR4

start http-sample: GOMAXPROCS=4 go run ./examples/codes/http-sample/server.go

**Comment by alpha-baby** [comment]
> 感觉可以，需要试一下 http 的性能有没有影响，用 example 里的 demo 和 wrk 简单测一下 post 上来就行，
> 如果差不多的话下版本合并~

test environment： MacBook Pro (16-inch, 2019) 2.6 GHz 六核Intel Core i7 16 GB 2667 MHz DDR4

start http-sample: `GOMAXPROCS=4 go run ./examples/codes/http-sample/server.go`

### stress result without proxy:

```
$ wrk -t4 -c500 -d30s --latency http://127.0.0.1:8080/test-for-example-http
Running 30s test @ http://127.0.0.1:8080/test-for-example-http
  4 threads and 500 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     2.48ms  744.02us  22.21ms   82.53%
    Req/Sec    25.16k     4.25k   33.36k    75.75%
  Latency Distribution
     50%    2.33ms
     75%    2.78ms
     90%    3.29ms
     99%    5.07ms
  3003886 requests in 30.01s, 1.41GB read
  Socket errors: connect 251, read 91, write 0, timeout 0
Requests/sec: 100110.17
Transfer/sec:     48.02MB
``` 

### strat mosn
```bash
$ make build-local && GOMAXPROCS=4 ./build/bundles/v0.18.0/binary/mosnd -c ./examples/codes/http-sample/server_config.json
```

### stress result with proxy using mosn 

**before optimization**

```
$ wrk -t4 -c500 -d30s --latency http://127.0.0.1:2046/test-for-example-http
Running 30s test @ http://127.0.0.1:2046/test-for-example-http
  4 threads and 500 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     8.20ms    1.80ms  31.72ms   81.12%
    Req/Sec     7.62k   738.27     9.66k    71.17%
  Latency Distribution
     50%    7.89ms
     75%    8.77ms
     90%   10.20ms
     99%   14.79ms
  909566 requests in 30.01s, 452.80MB read
  Socket errors: connect 251, read 98, write 0, timeout 0
Requests/sec:  30304.52
Transfer/sec:     15.09MB
```

**after optimization**

```
$ wrk -t4 -c500 -d30s --latency http://127.0.0.1:2046/test-for-example-http
Running 30s test @ http://127.0.0.1:2046/test-for-example-http
  4 threads and 500 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     7.91ms    1.80ms  46.97ms   79.42%
    Req/Sec     7.90k     0.87k   10.77k    72.58%
  Latency Distribution
     50%    7.67ms
     75%    8.56ms
     90%    9.90ms
     99%   13.85ms
  943180 requests in 30.01s, 469.53MB read
  Socket errors: connect 251, read 99, write 0, timeout 0
Requests/sec:  31431.60
Transfer/sec:     15.65M
```

File: pkg/stream/http/stream.go [hunk 1]
```diff
@@ -348,6 +350,7 @@ func (conn *clientStreamConnection) CheckReasonError(connected bool, event api.C
 
 func (conn *clientStreamConnection) Reset(reason types.StreamResetReason) {
 	close(conn.bufChan)
```
Comment (cch123) [inline pkg/stream/http/stream.go:352]: bufChan 和 endRead 的顺序有问题么

**Comment by cch123** [comment]
LGTM

File: pkg/stream/http/stream.go [hunk 2]
```diff
@@ -540,6 +545,9 @@ func (conn *serverStreamConnection) CheckReasonError(connected bool, event api.C
 
 func (conn *serverStreamConnection) Reset(reason types.StreamResetReason) {
```
[the remark this answers is not in the document]
Comment (alpha-baby) [reply pkg/stream/http/stream.go:546]: `Reset()` function parameter is `types.StreamResetReason`,

[the remark this answers is not in the document]
Comment (alpha-baby) [reply pkg/stream/http/stream.go:546]: `OnEvent(event api.ConnectionEvent)` function parameter is `api.ConnectionEvent`
how to transform？

[the remark this answers is not in the document]
Comment (wangfakang) [reply pkg/stream/http/stream.go:546]: > `OnEvent(event api.ConnectionEvent)` function parameter is `api.ConnectionEvent`
> how to transform？

Refer to the function of `checkreasonerror`.

[the remark this answers is not in the document]
Comment (alpha-baby) [reply pkg/stream/http/stream.go]: please look the latest commit

File: pkg/stream/http/stream.go [hunk 3]
```diff
@@ -410,8 +414,14 @@ func newServerStreamConnection(ctx context.Context, connection api.Connection,
 
 func (conn *serverStreamConnection) OnEvent(event api.ConnectionEvent) {
 	if event.IsClose() {
-		close(conn.bufChan)
-		close(conn.connClosed)
+		reason := types.StreamConnectionFailed
+		switch event {
+		case api.RemoteClose:
+			reason = types.UpstreamReset
```
Comment (nejisama) [inline pkg/stream/http/stream.go:420]: 这个无效的赋值处理一下

Comment (alpha-baby) [reply pkg/stream/http/stream.go:420]: fix

**Review by wangfakang — APPROVED** [review]
LGTM.

**Review by nejisama — APPROVED** [review]
